### PR TITLE
Add request response logging via fluentd

### PR DESCRIPTION
**Pull request by terryyylim**
**What this PR does / why we need it**:
This PR introduces the option to enable request/response logging to an external fluentd service. This is an addition to the existing Audit logging capabilities which logs to the console.

**Which issue(s) this PR fixes**:

N/A

**Does this PR introduce a user-facing change?**:

```release-note
Users can now enable forwarding of logs to an external fluentd service that's been setup.
```

### CODE REVIEW

File: common/pom.xml [hunk 1]
```diff
@@ -114,5 +114,10 @@
             <version>4.0</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.fluentd</groupId>
```
Comment (woop) [inline common/pom.xml:118]: Do we really need to pull this into all modules? Can this be moved to serving only?

File: common/src/main/java/feast/common/logging/AuditLogger.java [hunk 2]
```diff
@@ -37,8 +41,10 @@
 @Component
 public class AuditLogger {
   private static final Marker AUDIT_MARKER = MarkerFactory.getMarker("AUDIT_MARK");
+  private static FluentLogger FLUENT_LOGGER;
```
Comment (woop) [inline common/src/main/java/feast/common/logging/AuditLogger.java:44]: Why is `FLUENT_LOGGER` capitalized?

File: common/src/main/java/feast/common/logging/AuditLogger.java [hunk 3]
```diff
@@ -61,6 +68,23 @@ public static void logMessage(Level level, MessageAuditLogEntry.Builder entryBui
             .setComponent(buildProperties.getArtifact())
             .setVersion(buildProperties.getVersion())
             .build());
+    Map<String, Object> fluentDLogs = new HashMap<>();
```
Comment (woop) [inline common/src/main/java/feast/common/logging/AuditLogger.java:71]: Nitpick, but I think I prefer `fluentdLogs` so that we dont have two capital letters after each other.

File: common/src/main/java/feast/common/logging/AuditLogger.java [hunk 4]
```diff
@@ -61,6 +68,23 @@ public static void logMessage(Level level, MessageAuditLogEntry.Builder entryBui
...
+
+    if (logForwarderProperties.isEnabled()) {
+      FLUENT_LOGGER =
+          FluentLogger.getLogger(
```
Comment (woop) [inline common/src/main/java/feast/common/logging/AuditLogger.java:75]: `To create Fluent Logger instances, users need to invoke getLogger method in FluentLogger class like org.slf4j, org.log4j logging libraries. The method should be called only once`

Does this mean you are creating a new logger every time? Why did you create a static instance?

File: common/src/main/java/feast/common/logging/AuditLogger.java [hunk 5]
```diff
@@ -61,6 +68,23 @@ public static void logMessage(Level level, MessageAuditLogEntry.Builder entryBui
...
+              "feast",
+              logForwarderProperties.getFluentdHost(),
+              logForwarderProperties.getFluentdPort());
+      fluentDLogs.put("id", entryBuilder.build().getId());
+      fluentDLogs.put("service", entryBuilder.build().getService());
+      fluentDLogs.put("status_code", entryBuilder.build().getStatusCode());
+      fluentDLogs.put("method", entryBuilder.build().getMethod());
+      fluentDLogs.put("request", entryBuilder.build().getRequest());
```
Comment (woop) [inline common/src/main/java/feast/common/logging/AuditLogger.java:83]: Request and response should be in Json, just like the existing audit logger. Otherwise its very hard to parse in BQ.

Comment (woop) [reply common/src/main/java/feast/common/logging/AuditLogger.java:83]: https://github.com/feast-dev/feast/blob/21509be8717674575db6edb6f67378b40043e1e3/common/src/main/java/feast/common/logging/entry/MessageAuditLogEntry.java#L97

File: common/src/main/java/feast/common/logging/config/LoggingProperties.java [hunk 6]
```diff
@@ -24,6 +24,20 @@
 @Setter
 public class LoggingProperties {
   @NotNull private AuditLogProperties audit;
+  @NotNull private LogForwarderProperties forwarder;
```
Comment (woop) [inline common/src/main/java/feast/common/logging/config/LoggingProperties.java:27]: `logForwarderProperties`

File: common/src/main/java/feast/common/logging/config/LoggingProperties.java [hunk 7]
```diff
@@ -24,6 +24,20 @@
...
+
+  @Getter
+  @Setter
+  public static class LogForwarderProperties {
+    // Whether to enable/disable (request/response) log forwarding entirely.
```
Comment (woop) [inline common/src/main/java/feast/common/logging/config/LoggingProperties.java:32]: If you enable this, and disable audit logging, what happens?

File: common/pom.xml (hunk 1 above)
Comment (terryyylim) [reply common/pom.xml:118]: AuditLogger is used in both Core and Serving, and it exists in common module, so I think that's not possible.

File: common/src/main/java/feast/common/logging/AuditLogger.java (hunk 3 above)
Comment (terryyylim) [reply common/src/main/java/feast/common/logging/AuditLogger.java]: Updated.

File: common/src/main/java/feast/common/logging/AuditLogger.java (hunk 5 above)
Comment (terryyylim) [reply common/src/main/java/feast/common/logging/AuditLogger.java]: Updated.

File: common/src/main/java/feast/common/logging/config/LoggingProperties.java (hunk 6 above)
Comment (terryyylim) [reply common/src/main/java/feast/common/logging/config/LoggingProperties.java]: Updated.

File: common/src/main/java/feast/common/logging/AuditLogger.java (hunk 4 above)
Comment (terryyylim) [reply common/src/main/java/feast/common/logging/AuditLogger.java]: Updated, creating through constructor now.

File: common/src/main/java/feast/common/logging/interceptors/GrpcMessageInterceptor.java [hunk 8]
```diff
@@ -86,6 +87,7 @@ public void sendMessage(RespT message) {
             entryBuilder.setResponse((Message) message);
           }
 
+          @SneakyThrows
```
Comment (woop) [inline common/src/main/java/feast/common/logging/interceptors/GrpcMessageInterceptor.java:90]: Please don't use this lombok annotation. We are only using Lombok for getters and setters.

File: core/src/main/resources/application.yml [hunk 9]
```diff
@@ -51,6 +51,12 @@ feast:
     disableRestControllerAuth: false
 
   logging:
+    # Logging forwarder currently provides a machine readable structured JSON log to an
+    # external fluentd service that can give better insight into what is happening in Feast.
+    logForwarderProperties:
+      enabled: false
```
Comment (woop) [inline core/src/main/resources/application.yml:57]: What configuration is necessary in order to log using the forwarder but **NOT** log to the console?

File: common/src/main/java/feast/common/logging/AuditLogger.java [hunk 10]
```diff
@@ -47,6 +53,13 @@ public AuditLogger(LoggingProperties loggingProperties, BuildProperties buildPro
     // This allows us to use the dependencies in the AuditLogger's static methods
     AuditLogger.properties = loggingProperties.getAudit();
     AuditLogger.buildProperties = buildProperties;
+    if (AuditLogger.properties.getMessageLogging().isEnabled()) {
```
Comment (pyalex) [inline common/src/main/java/feast/common/logging/AuditLogger.java:56]: can `properties.getMessageLogging()` return null?

File: common/src/main/java/feast/common/logging/AuditLogger.java [hunk 11]
```diff
@@ -113,28 +126,49 @@ private static void log(Level level, AuditLogEntry entry) {
...
-        break;
-      case WARN:
-        log.warn(AUDIT_MARKER, entryJSON);
-        break;
-      case ERROR:
-        log.error(AUDIT_MARKER, entryJSON);
-        break;
+    // Either forward log to logging layer or log to console
+    Map<String, Object> fluentdLogs = new HashMap<>();
+    if (properties.getMessageLogging().getDestination().equals("fluentd")) {
```
Comment (pyalex) [inline common/src/main/java/feast/common/logging/AuditLogger.java:135]: maybe `fluentd` to constant?

File: common/src/main/java/feast/common/logging/AuditLogger.java [hunk 12]
```diff
@@ -113,28 +126,49 @@ private static void log(Level level, AuditLogEntry entry) {
...
+      MessageAuditLogEntry messageAuditLogEntry = (MessageAuditLogEntry) entry;
+      try {
+        fluentdLogs.put("id", messageAuditLogEntry.getId());
```
Comment (pyalex) [inline common/src/main/java/feast/common/logging/AuditLogger.java:138]: which one of this throws exception? do we need to put all of them into `try`?

File: common/src/main/java/feast/common/logging/AuditLogger.java [hunk 13]
```diff
@@ -113,28 +126,49 @@ private static void log(Level level, AuditLogEntry entry) {
...
-        log.info(AUDIT_MARKER, entryJSON);
-        break;
-      case WARN:
-        log.warn(AUDIT_MARKER, entryJSON);
-        break;
-      case ERROR:
-        log.error(AUDIT_MARKER, entryJSON);
-        break;
+    // Either forward log to logging layer or log to console
+    Map<String, Object> fluentdLogs = new HashMap<>();
```
Comment (pyalex) [inline common/src/main/java/feast/common/logging/AuditLogger.java:134]: I suggest to replace this with `ImmutableMap` since there's no need for mutability and code would contain less of `fluentdLogs.put`

File: common/src/main/java/feast/common/logging/AuditLogger.java [hunk 14]
```diff
@@ -113,28 +130,52 @@ private static void log(Level level, AuditLogEntry entry) {
...
+      try {
+        fluentdLogs.put("request", JsonFormat.printer().print(messageAuditLogEntry.getRequest()));
+        fluentdLogs.put("response", JsonFormat.printer().print(messageAuditLogEntry.getResponse()));
+      } catch (InvalidProtocolBufferException e) {
+        throw io.grpc.Status.NOT_FOUND
+            .withDescription("Unable to forward logs to logging service")
+            .withCause(e)
+            .asRuntimeException();
+      }
+      // tag can be used as "service" to split rr logs into separate tables
```
Comment (pyalex) [inline common/src/main/java/feast/common/logging/AuditLogger.java:157]: not sure this comment related here

File: common/src/main/java/feast/common/logging/AuditLogger.java [hunk 15]
```diff
@@ -36,7 +42,10 @@
 @Slf4j
 @Component
 public class AuditLogger {
+  private static final String DESTINATION = "fluentd";
+  private static final String DEFAULT_NAME = "feast-0-0";
```
Comment (pyalex) [inline common/src/main/java/feast/common/logging/AuditLogger.java:46]: probably `DEFAULT_RELEASE_NAME`

File: common/src/main/java/feast/common/logging/AuditLogger.java [hunk 16]
```diff
@@ -113,28 +130,52 @@ private static void log(Level level, AuditLogEntry entry) {
...
+      fluentdLogs.put("id", messageAuditLogEntry.getId());
+      fluentdLogs.put("service", messageAuditLogEntry.getService());
+      fluentdLogs.put("status_code", messageAuditLogEntry.getStatusCode());
+      fluentdLogs.put("method", messageAuditLogEntry.getMethod());
+      fluentdLogs.put("release_name", releaseName);
+      try {
+        fluentdLogs.put("request", JsonFormat.printer().print(messageAuditLogEntry.getRequest()));
+        fluentdLogs.put("response", JsonFormat.printer().print(messageAuditLogEntry.getResponse()));
+      } catch (InvalidProtocolBufferException e) {
+        throw io.grpc.Status.NOT_FOUND
```
Comment (pyalex) [inline common/src/main/java/feast/common/logging/AuditLogger.java:153]: NOT_FOUND and description are not really represent the cause of the exception that may happen here

File: common/src/main/java/feast/common/logging/AuditLogger.java [hunk 17]
```diff
@@ -113,28 +130,52 @@ private static void log(Level level, AuditLogEntry entry) {
       return;
     }
     if (entry.getKind().equals(AuditLogEntryKind.MESSAGE)
-        && !properties.isMessageLoggingEnabled()) {
+        && !properties.getMessageLogging().isEnabled()) {
```
Comment (pyalex) [inline common/src/main/java/feast/common/logging/AuditLogger.java:133]: similar issue with NPE as on line 59

File: common/src/main/java/feast/common/logging/AuditLogger.java (hunk 16 above)
Comment (terryyylim) [reply common/src/main/java/feast/common/logging/AuditLogger.java]: Updated description and status type.

File: common/src/main/java/feast/common/logging/config/LoggingProperties.java [hunk 18]
```diff
@@ -31,7 +31,22 @@
...
+    private MessageLogging messageLogging;
+
+    @Getter
+    @Setter
+    public static class MessageLogging {
+      // Whether to enable/disable message level (ie request/response) audit logging.
+      private boolean enabled;
+
+      // Whether to log to console or fluentd
+      private String destination;
```
Comment (woop) [inline common/src/main/java/feast/common/logging/config/LoggingProperties.java:43]: Can we use the `OneOf` annotation here for safety?

File: common/src/main/java/feast/common/logging/AuditLogger.java [hunk 19]
```diff
@@ -112,24 +129,47 @@ private static void log(Level level, AuditLogEntry entry) {
...
-        log.info(AUDIT_MARKER, entryJSON);
-        break;
-      case WARN:
-        log.warn(AUDIT_MARKER, entryJSON);
-        break;
-      case ERROR:
-        log.error(AUDIT_MARKER, entryJSON);
-        break;
+    // Either forward log to logging layer or log to console
+    if (properties.getMessageLogging().getDestination().equals(FLUENTD_DESTINATION)) {
```
Comment (woop) [inline common/src/main/java/feast/common/logging/AuditLogger.java:133]: What prevents this code from executing if message Logging is disabled?

Comment (terryyylim) [reply common/src/main/java/feast/common/logging/AuditLogger.java:133]: That's handled [here](https://github.com/feast-dev/feast/blob/master/common/src/main/java/feast/common/logging/interceptors/GrpcMessageInterceptor.java#L73) from #940.

Comment (woop) [reply common/src/main/java/feast/common/logging/AuditLogger.java:133]: By the way, as a convention we try and instantiate variables from getters in order to make code more readable. So something like
```
String messageLogDestination = properties.getMessageLogging().getDestination();
if (message.equals(FLUENTD_DESTINATION)) {

}
```

**Review by woop — APPROVED** [review]
